[case testTypeClause]
from typing import Any

from sqlalchemy import Integer
from sqlalchemy.sql.elements import TypeClause

clause = TypeClause(Integer())
reveal_type(clause.type)  # N: Revealed type is 'sqlalchemy.sql.type_api.TypeEngine[builtins.int*]'
[out]

[case testNull]
from typing import Any

from sqlalchemy.sql.elements import Null

null = Null()
reveal_type(null.type)  # N: Revealed type is 'sqlalchemy.sql.sqltypes.NullType'
[out]

[case testFalse_]
from typing import Any

from sqlalchemy.sql.elements import False_

false_ = False_()
reveal_type(false_.type)  # N: Revealed type is 'sqlalchemy.sql.sqltypes.Boolean'
[out]

[case testTrue_]
from typing import Any

from sqlalchemy.sql.elements import True_

true_ = True_()
reveal_type(true_.type)  # N: Revealed type is 'sqlalchemy.sql.sqltypes.Boolean'
[out]

[case testTupleInferred]
from typing import Any

from sqlalchemy import Boolean
from sqlalchemy.sql.elements import Tuple, Null, False_

tp = Tuple(Null(), False_())
reveal_type(tp.type)  # N: Revealed type is 'sqlalchemy.sql.type_api.TypeEngine[None]'
[out]

[case testTupleExplicit]
from typing import Any

from sqlalchemy import Boolean
from sqlalchemy.sql.elements import Tuple, Null, False_

tp = Tuple(Null(), False_(), type_=Boolean)
reveal_type(tp.type)  # N: Revealed type is 'sqlalchemy.sql.type_api.TypeEngine[builtins.bool*]'
[out]

[case testTupleExplicitInstance]
from typing import Any

from sqlalchemy import Boolean
from sqlalchemy.sql.elements import Tuple, Null, False_

tp = Tuple(Null(), False_(), type_=Boolean())
reveal_type(tp.type)  # N: Revealed type is 'sqlalchemy.sql.type_api.TypeEngine[builtins.bool*]'
[out]

[case testCase]
from typing import Any, List

from sqlalchemy import Column, case

column: Column[Any]
lst: List[Any]
reveal_type(case(value=column, whens={item: i for i, item in enumerate(lst)}))  # N: Revealed type is 'sqlalchemy.sql.elements.Case[builtins.int*]'
[out]

[case testCast]
from typing import Any, Set

from sqlalchemy.ext.declarative import declarative_base
from sqlalchemy import Column, String, ARRAY, cast

Base: Any = declarative_base()

class Model(Base):
    __tablename__ = 'users'
    tags = Column(ARRAY(String(16)), nullable=False)

tags: Set[str] = set()
reveal_type(cast(tags, Model.tags.type))  # N: Revealed type is 'sqlalchemy.sql.elements.Cast[builtins.list*[builtins.str]]'
[out]

